ostd/src/exception/handler.rs

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! General trap handling in OSTD.
+
+use core::sync::atomic::{AtomicBool, Ordering};
+
+use align_ext::AlignExt;
+use log::debug;
+#[cfg(feature = "intel_tdx")]
+use tdx_guest::tdcall;
+use trapframe::TrapFrame;
+
+#[cfg(feature = "intel_tdx")]
+use crate::arch::tdx_guest::handle_virtual_exception;
+use crate::{
+    arch::{ex_table::ExTable, irq::IRQ_LIST},
+    cpu::{CpuException, CpuExceptionInfo, CpuExceptionType, PageFaultErrorCode, UserContext},
+    cpu_local,
+    mm::{
+        kspace::{KERNEL_PAGE_TABLE, LINEAR_MAPPING_BASE_VADDR, LINEAR_MAPPING_VADDR_RANGE},
+        page_prop::{CachePolicy, PageProperty},
+        PageFlags, PrivilegedPageFlags as PrivFlags, MAX_USERSPACE_VADDR, PAGE_SIZE,
+    },
+    task::current_task,
+    user::{ReturnReason, UserContextApiInternal},
+};
+
+cpu_local! {
+    static IN_INTERRUPT_CONTEXT: AtomicBool = AtomicBool::new(false);
+}
+
+/// Returns true if this function is called within the context of an IRQ handler
+/// and the IRQ occurs while the CPU is executing in the kernel mode.
+/// Otherwise, it returns false.
+///
+/// FIXME: For ISAs that supports re-entrant interrupts, we may need to record nested level here.
+pub fn in_interrupt_context() -> bool {
+    IN_INTERRUPT_CONTEXT.load(Ordering::Acquire)
+}
+
+// For x86 CPUs, interrupts are not re-entrant. Local interrupts will be disabled when
+// an interrupt handler is called (unless interrupts are re-enabled in an interrupt handler).
+fn re_enable_interrupts() {
+    #[cfg(target_arch = "x86_64")]
+    {
+        debug_assert!(!crate::arch::irq::is_local_enabled());
+        crate::arch::irq::enable_local();
+    }
+}
+
+/// Handles trap originated from the user mode.
+///
+/// It returns whether we can immediately return to the user mode without any
+/// further handling. If so, the return value is `None`. Otherwise it returns
+/// the reason why the user traps into the kernel mode. The caller can use it
+/// to decide how to further handle the trap.
+pub(crate) fn user_mode_exception_handler(ctx: &mut UserContext) -> Option<ReturnReason> {
+    let trap_frame = ctx.as_trap_frame();
+    let exception = CpuException::from_num(trap_frame.trap_num as u16);
+    let reaction = match exception {
+        #[cfg(feature = "intel_tdx")]
+        CpuException::VirtualizationException => {
+            let ve_info = tdcall::get_veinfo().expect("#VE handler: fail to get VE info\n");
+            handle_virtual_exception(ctx.general_regs_mut(), &ve_info);
+            None
+        }
+        CpuException::Syscall => Some(ReturnReason::UserSyscall),
+        _ => match exception.get_type() {
+            CpuExceptionType::FaultOrTrap | CpuExceptionType::Fault | CpuExceptionType::Trap => {
+                Some(ReturnReason::UserException)
+            }
+            CpuExceptionType::Interrupt => {
+                handle_interrupt(&trap_frame);
```

Comment:
Actually, the CPU exception with interrupt type in x86 is called Non-maskable Interrupt (NMI). Hardware may trigger NMI when **unrecoverable or critical situations** happen. For now, we can panic here since we have no idea how to handle NMI.